### PR TITLE
Added PackagePublisher tool

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: BuildTools.sln [hunk 1]
```diff
@@ -15,34 +15,36 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{60A938B2-D95A-403C-AA7A-3683AD64DFA0}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NuGetPackageVerifier", "src\NuGetPackageVerifier\NuGetPackageVerifier.csproj", "{657AFF5E-164E-493D-8501-8026B7C20808}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "NuGetPackageVerifier", "src\NuGetPackageVerifier\NuGetPackageVerifier.csproj", "{657AFF5E-164E-493D-8501-8026B7C20808}"
```
Comment (kichalla) [inline BuildTools.sln]: I will revert these changes

**Comment by natemcmaster** [comment]
I'm missing context here. Why do we need this? Have we considered using `dotnet nuget push` instead?

**Comment by kichalla** [comment]
Ah sure, We would like to use this tool to have retry logic as recently we have been seeing failures when uploading packages.

**Comment by natemcmaster** [comment]
Thanks for clarifying. FWIW code looks fine. Just seems like overkill to create our own console tool when a few lines of bash/powershell with a while loop would provide the same functionality.

File: src/PackagePublisher/Log.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace PackagePublisher
+{
+    public class Log
```
Comment (pranavkm) [inline src/PackagePublisher/Log.cs]: Don't need this.

File: src/PackagePublisher/Program.cs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        private static int _maxRetryCount = 10;
+
+        public static int Main(string[] args)
+        {
+            var app = new CommandLineApplication();
+            _packagesDirectory = app.Option(
+                "-d|--directory",
+                "The directory containing .nupkg file(s)",
+                CommandOptionType.SingleValue);
+            _packagePath = app.Option(
```
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: I thought we landed on leaving Coherence along and just creating a tool that pushed packages in a dir?

File: src/PackagePublisher/Program.cs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                {
+                    if (!Directory.Exists(packageDir))
+                    {
+                        Log.WriteError($"The supplied directory '{packageDir}' does not exist.");
+                        return 1;
+                    }
+                    else
+                    {
+                        _packages = new DirectoryInfo(packageDir)
+                        .EnumerateFiles("*.nupkg", SearchOption.AllDirectories)
```
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: tab

File: src/PackagePublisher/Program.cs [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+                        .ProcessPackages();
+                    }
+                }
+                else
+                {
+                    var files = new List<FileInfo>();
+                    foreach (var filePath in _packagePath.Values)
```
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: Let's remove this. I don't think we have a good use case for this as yet.

File: src/PackagePublisher/Program.cs [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            {
+                Log.WriteError("Feed is required.");
+                return false;
+            }
+
+            _apiKey = Environment.GetEnvironmentVariable(ApiKeyEnvironmentVariableName);
+            if (string.IsNullOrEmpty(_apiKey))
+            {
+                Log.WriteError("Api key is required to publish to the feed. Set the environment " +
+                    $"variable {ApiKeyEnvironmentVariableName} before inovking this tool again.");
```
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: typo `invoking`

File: src/PackagePublisher/Program.cs [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+                            Identity = reader.GetIdentity(),
+                            PackagePath = fileInfo.FullName
+                        };
+                    }
+                });
+        }
+
+        private static async Task PublishToFeedAsync()
+        {
+            using (var semaphore = new SemaphoreSlim(initialCount: 4, maxCount: 4))
```
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: We can do a second iteration where we can avoid publishing packages we've previously pushed. For now, you don't need the call to `ProcessPackages`

File: src/PackagePublisher/Program.cs (hunk 4 above)
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: `.Where(file => !file.Path.EndsWith(".symbols.nupkg", StringComparison.OrdinalIgnoreCase);`

File: src/PackagePublisher/Program.cs [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+using NuGet.Common;
+using NuGet.Packaging;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace PackagePublisher
+{
+    public static class Program
+    {
+        private const string ApiKeyEnvironmentVariableName = "PACKAGEPUBLISHER_APIKEY";
```
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: `APIKey`

File: src/PackagePublisher/PackagePublisher.csproj [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  <Import Project="..\..\build\common.props" />
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp1.1</TargetFrameworks>
+    <OutputType>exe</OutputType>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.CommandLineUtils" Version="1.0.1" />
+    <PackageReference Include="NuGet.Packaging" Version="3.5.0" />
+    <PackageReference Include="NuGet.Protocol.Core.v3" Version="3.5.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
```
Comment (ajaybhargavb) [inline src/PackagePublisher/PackagePublisher.csproj]: $(JsonNetVersion)

File: src/PackagePublisher/PackagePublisher.csproj [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="..\..\build\common.props" />
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp1.1</TargetFrameworks>
+    <OutputType>exe</OutputType>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.CommandLineUtils" Version="1.0.1" />
+    <PackageReference Include="NuGet.Packaging" Version="3.5.0" />
```
Comment (ajaybhargavb) [inline src/PackagePublisher/PackagePublisher.csproj]: $(NuGetPackagesVersion)

File: src/PackagePublisher/Program.cs [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                "-c|--retryCount",
+                $"Max retry count to publish a package. Defaults to {_maxRetryCount}",
+                CommandOptionType.SingleValue);
+            var help = app.Option("-h|--help", "Show help", CommandOptionType.NoValue);
+
+            app.OnExecute(() =>
+            {
+                if (help.HasValue())
+                {
+                    app.ShowHelp();
```
Comment (ajaybhargavb) [inline src/PackagePublisher/Program.cs]: Curious, where does this call into?

File: src/PackagePublisher/Program.cs [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+using NuGet.Common;
+using NuGet.Packaging;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace PackagePublisher
+{
+    public static class Program
+    {
+        private const string ApiKeyEnvironmentVariableName = "PACKAGEPUBLISHER_APIKey";
```
Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: Nope. Just `APIKey`. That's what we use now and I'd rather not change it.

File: src/PackagePublisher/Program.cs (hunk 11 above)
Comment (kichalla) [inline src/PackagePublisher/Program.cs]: The commandlineapplication object has a standard way to show up the help that you would normally see when you do, let's say `dotnet restore --help`

**Comment by kichalla** [comment]
e0f6fa4233c6fbaf6f1d69d98a890252e55ae673

File: src/PackagePublisher/Program.cs [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+                            Identity = reader.GetIdentity(),
+                            PackagePath = fileInfo.FullName
+                        };
+                    }
+                });
+        }
+
+        private static async Task PublishToFeedAsync()
+        {
+            using (var semaphore = new SemaphoreSlim(initialCount: 4, maxCount: 4))
```
Comment (mikeharder) [inline src/PackagePublisher/Program.cs]: Is there a specific reason uploads are limited to 4 in parallel?  Would it be faster if this was increased?

Comment (pranavkm) [inline src/PackagePublisher/Program.cs]: We've historically seen timeouts increase when we bumped up this number. Although that was when we were publishing to `www.myget.org` as opposed to `dotnet.myget.org`.

Comment (mikeharder) [inline src/PackagePublisher/Program.cs]: Seems reasonable.  @kichalla: Can you please add a comment to the source code explaining this?

**Pull request closed without merging**